torchprime/torch_xla_models/experimental/custom_kernel.py

```diff
@@ -0,0 +1,471 @@
+    global_use_fused_bwd_kernel = config.sa_use_fused_bwd_kernel
+    global_q_layout = config.sa_q_layout
+    global_k_layout = config.sa_k_layout
+    global_v_layout = config.sa_v_layout
+    shard_map = shard_map.shard_map
+
+    @functools.partial(
+      shard_map,
+      mesh=mesh,
+      in_specs=(
+        axis_names,
+        axis_names,
+        axis_names,
+        segment_axis_names,
+      ),
+      out_specs=axis_names,
+      check_rep=False,
+    )
+    def wrap_flash_attention(query, key, value, decoder_segment_ids):
+      if decoder_segment_ids is not None:
+        assert (
+          query.shape[2] == decoder_segment_ids.q.shape[1]
+        ), "Sharding along sequence dimension not allowed in tpu kernel attention"
+      block_sizes = splash_attention_kernel.BlockSizes(
+        block_q=min(global_block_q, query.shape[2]),
+        block_kv=min(global_block_kv, key.shape[2]),
+        block_kv_compute=min(global_block_kv_compute, key.shape[2]),
+        block_q_dkv=min(global_block_q_dkv, query.shape[2]),
+        block_kv_dkv=min(global_block_kv_dkv, key.shape[2]),
+        block_kv_dkv_compute=min(global_block_kv_dkv_compute, query.shape[2]),
+        block_q_dq=None
+        if global_use_fused_bwd_kernel
+        else min(global_block_q_dq, query.shape[2]),
+        block_kv_dq=None
+        if global_use_fused_bwd_kernel
+        else min(global_block_kv_dq, query.shape[2]),
+        use_fused_bwd_kernel=global_use_fused_bwd_kernel,
+        q_layout=splash_attention_kernel.QKVLayout[global_q_layout],
+        k_layout=splash_attention_kernel.QKVLayout[global_k_layout],
+        v_layout=splash_attention_kernel.QKVLayout[global_v_layout],
+      )
+
+      mask = splash_attention_mask.CausalMask(shape=(query.shape[2], query.shape[2]))
+
+      # Apply local masking if local sliding attention is enabled.
+      if config.AttentionType_LOCAL_SLIDING:
+        if config.SLIDE_WINDOW_SIZE is None:
+          raise ValueError(
+            "Sliding_window_size must be set if Local Sliding attention type"
+          )
+        mask &= splash_attention_mask.LocalMask(
+          shape=(query.shape[2], query.shape[2]),
+          window_size=(config.SLIDE_WINDOW_SIZE, config.SLIDE_WINDOW_SIZE),
+          offset=0,
+        )
+
+      # Create multi-head mask
+      multi_head_mask = splash_attention_mask.MultiHeadMask(
+        masks=(mask,) * query.shape[1]
+      )
+      splash_kernel = splash_attention_kernel.make_splash_mha(
+        mask=multi_head_mask,
+        head_shards=1,
+        q_seq_shards=1,
+        block_sizes=block_sizes,
+        attn_logits_soft_cap=attn_logits_soft_cap,
+      )
+      return jax.vmap(splash_kernel)(query, key, value, segment_ids=decoder_segment_ids)
+
+    devices_in_data_fsdp = mesh.shape["data"] * mesh.shape["fsdp"]
+    assert (query.shape[0] / devices_in_data_fsdp).is_integer(), (
+      "Batch dimension should be shardable among the devices in data and fsdp" " axis"
```

Comment:
nit: ` "Batch dimension should be shardable among the devices in data and fsdp axis"`